docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,725 @@
+`b2 => d2` to make the optional trigger work by creating an unsatisfied
+dependency for the unused pipelined)
+
+## Another Example
+
+Intent:
+- if `a` succeeds, run `b`
+- if `a` fails, run `r1 => b => r2`
+
+This is a pathological case for Cylc 7 and 8 because `b` appears with different
+dependencies in alternate branches, but for the reasons given above both
+branches contribute their `b` dependencies to the same task.
+
+Cylc 9 will handle this properly and intuitively, but we can't do it yet:
+```
+a? => {b}
+a:fail? => {r1 => b => r2}
+```
+...the lower sub-graph does not add its dependencies to `b` unless `a` fails.
+
+Cylc 7 and 8 workaround: use different logical tasks (with different
+dependencies) for the same job in each branch:
+```
+a? => b1
+a:fail? => r1 => b2 => r2
+
+[runtime]
+   [[B]]
+   [[b1, b2]]
+       inherit = B
+```
+
+# Appendix: Output Completion Expressions?
+
+A possible follow-up to this proposal.
+
+In an example above we used an artificial dependency (possibly triggering a
+dummy task) to deliberately generate a problem that can stall the workflow
+if neither branch runs in an alternate branch case:
+```
+a:x? => b1
+a:y? => b2
+b1 | b2 => c  # join
+a => c # ARTIFICIAL DEPENDENCY
+```
+
+The artificial dependency causes an unsatisfied `c` to be spawned even if
+neither branch runs.
+
+However, this isn't an ideal way to alert users to the problem
+- it might not occur to users to use an artificial dependency like this
+- unsatisfied prerequisites can't be definitively identified as errors
+  until/unless the workflow stalls
+- unsatisfied prerequisites are an indirect problem indicator: they do not
+  single out the task that actually caused the problem (which is that `a`
+should have generated either `x` or `y`)
+
+So maybe we can (correctly) identify `a` as incomplete instead:
+
+```
+a:succeed(x|y) => c
+
+# or in abbreviated success form:
+a(x|y) => c
+
+# or if direct triggering of c is not needed:
+a(x|y)
+```
+This (or some variation on the suggested syntax) says that `a` is required to
+succeed with at least one of `x` and `y` completed; otherwise it is incomplete
+even if it reports success.
+
+## Caveat?
+
+What if the optional outputs are not from the same task?
+```
+a => b1? & b2?
+b1? | b2? => c  # join
+a => c # ARTIFICIAL DEPENDENCY
+```
+In this case the artificial dependency is still needed to ensure that either
+`b1` or `b2` succeed.
+
+However, this example is arguably a workflow design error.
+
+If a task generates multiple outputs it is possible to say whether or not at
+least one output is expected - that is down to the internal functioning of the
+task. And if at least one is expected but it doesn't happen, then the task is
+incomplete.
+
+But if `b1` and `b2` above are both success-optional independent tasks, then
+surely two failures at once is a perfectly possible normal outcome, in which
+case it would make sense for the joining dependency to not be triggered.
```

Comment:
It is a possible outcome but it may not be what the user requires. For example, `b1` and `b2` could be tasks that go and retrieve data from sources that may not always be available. `c` can run so long as at least  one of `b1` and `b2` succeed but the workflow can't complete if both fail. This seems a perfectly reasonable scenario to me.

Reply:
OK so you want the workflow to stall if both b1 and b2 fail, even though that is to be expected sometimes, and when it does  occur there would be nothing for `c` to do. To me that seems like a case where it makes perfect sense to shut down cleanly with nothing more to do, rather than stall.

If you want to be alerted to the fact that both b1 and b2 failed, then expecting `c` to be spawned as unsatisfied to cause an eventual stall seems a funny way to do it (Cylc 7 Stockholm syndrome? :grin: ). It would be better to do this:

```
b1? & b2? => c
b1:fail? & b2:fail? => alert
```

Reply:
Note that in the single task case: 
```
b:x => m
b:y => n
m | n => z
```
... I've been assuming that we can say, based on knowledge of what `b` does, that something has gone wrong if `b` does not complete one output or the other. Either `b` encountered an error but did report failure, or the workflow writer has misunderstood how `b` is supposed to behave.  In which case, having `x | y` as an output completion condition would allow us to flag `b` as incomplete.

If on the other hand not completing either x or y is a possible normal outcome of `b`, then we're in exactly the same position as your two-task example, and it wouldn't be appropriate to use the output completion condition in that case.

Reply:
(I'll update the text on this)

Reply:
> OK so you want the workflow to stall if both b1 and b2 fail, even though that is to be expected sometimes, and when it does occur there would be nothing for c to do. To me that seems like a case where it makes perfect sense to shut down cleanly with nothing more to do, rather than stall.

When we have unhandled task failures we stall to allow the user to intervene - the fact that the failure is unhandled doesn't necessarily mean it is unexpected. I view this example in the same way. It's not really important whether the situation (both tasks failing) is expected - what's important is whether the workflow is designed to handle the situation. If it isn't then we should stall to allow intervention.

(This point doesn't really affect the proposal so not a big deal if we view it slightly differently)

Reply:
> what's important is whether the workflow is designed to handle the situation

Bearing in mind this discussion relates to my proprosed "output completion condition" extension ... in the single task multiple output case, we can *know* that the task is always supposed to generate on output or the other, so if it doesn't generate either one that's a definite error.  But in the two-task case we know that it is quite possible to for both to fail, and I don't see why we should assume that's a stall-worthy problem rather than a possible normal outcome for which the joining task should not be spawned at all.

(But yes, we can't punt this discussion to the future as it won't be in the initial implementation if at all)